File: lib/board-update.ts
import type { NarrativePhase, NovelSettings, Persona } from "./types";
import { BOARD_UPDATE_PROMPTS } from "./prompts";

export type BoardUpdateRequest = {
  apiKey: string;
  model: string;
  phase: NarrativePhase;
  persona: Persona;
  settings: NovelSettings;
  userInstruction: string;
  assistantResponse: string;
};

export type BoardUpdateResult = {
  title?: string;
  concept?: string;
  characters?: string;
  worldView?: string;
  plot?: string;
  referenceLinks?: string;
  writingRules?: string;
};

export function buildBoardUpdateMessages(input: BoardUpdateRequest) {
  return [
    {
      role: "system",
      content: BOARD_UPDATE_PROMPTS.SYSTEM,
    },
    {
      role: "user",
      content: [
        `現在フェーズ: ${input.phase}`,
        "",
        "AI作家ペルソナ:",
        `${input.persona.name}: ${input.persona.style}`,
        "",
        "現在のボード:",
        `タイトル: ${input.settings.title}`,
        "",
        "コンセプト:",
        input.settings.concept || "未設定",
        "",
        "キャラクター設定:",
        input.settings.characters || "未設定",
        "",
        "世界観・ルール:",
        input.settings.worldView || "未設定",
        "",
        "プロット:",
        input.settings.plot || "未設定",
        "",
        "参照リンク・素材メモ:",
        input.settings.referenceLinks || "なし",
        "",
        "文体・本稿ルール:",
        input.settings.writingRules || "未設定",
        "",
        "最新のユーザー発言:",
        input.userInstruction,
        "",
        "最新のAI応答:",
        input.assistantResponse,
        "",
        BOARD_UPDATE_PROMPTS.USER_FOOTER,
      ].join("\n"),
    },
  ];
}

export function parseBoardUpdate(raw: string): BoardUpdateResult {
  const trimmed = raw.trim();
  const jsonText = trimmed.startsWith("{")
    ? trimmed
    : trimmed.match(/\{[\s\S]*\}/)?.[0] || "{}";
  const parsed = JSON.parse(jsonText) as BoardUpdateResult;

  return {
    title: typeof parsed.title === "string" ? parsed.title.trim() : undefined,
    concept: typeof parsed.concept === "string" ? parsed.concept.trim() : undefined,
    characters: typeof parsed.characters === "string" ? parsed.characters.trim() : undefined,
    worldView: typeof parsed.worldView === "string" ? parsed.worldView.trim() : undefined,
    plot: typeof parsed.plot === "string" ? parsed.plot.trim() : undefined,
    referenceLinks: typeof parsed.referenceLinks === "string" ? parsed.referenceLinks.trim() : undefined,
    writingRules: typeof parsed.writingRules === "string" ? parsed.writingRules.trim() : undefined,
  };
}

File: lib/board_update.rb
# frozen_string_literal: true

require "json"
require_relative "prompts"

module Aiwrite
  module BoardUpdate
    extend self

    # LLM 呼び出し用のペイロードを準備する
    def prepare_payload(input)
      persona  = input["persona"]  || {}
      settings = input["settings"] || {}

      user_content = [
        "現在フェーズ: #{input['phase']}",
        "",
        "AI作家ペルソナ:",
        "#{persona['name']}: #{persona['style']}",
        "",
        "現在のボード:",
        "タイトル: #{settings['title']}",
        "",
        "コンセプト:",
        settings["concept"].to_s.empty? ? "未設定" : settings["concept"],
        "",
        "キャラクター設定:",
        settings["characters"].to_s.empty? ? "未設定" : settings["characters"],
        "",
        "世界観・ルール:",
        settings["worldView"].to_s.empty? ? "未設定" : settings["worldView"],
        "",
        "プロット:",
        settings["plot"].to_s.empty? ? "未設定" : settings["plot"],
        "",
        "参照リンク・素材メモ:",
        settings["referenceLinks"].to_s.empty? ? "なし" : settings["referenceLinks"],
        "",
        "文体・本稿ルール:",
        settings["writingRules"].to_s.empty? ? "未設定" : settings["writingRules"],
        "",
        "最新のユーザー発言:",
        input["userInstruction"],
        "",
        "最新のAI応答:",
        input["assistantResponse"],
        "",
        Prompts::BOARD_UPDATE_PROMPTS[:user_footer]
      ].join("\n")

      {
        system_prompt: Prompts::BOARD_UPDATE_PROMPTS[:system],
        history: [],
        user_message: user_content
      }
    end

    # AI応答のJSON文字列をパースしてハッシュにする
    def parse(raw)
      trimmed = raw.strip
      json_text = if trimmed.start_with?("{")
                    trimmed
                  else
                    trimmed.match(/\{[\s\S]*\}/)&.to_s || "{}"
                  end

      parsed = JSON.parse(json_text)
      allowed = %w[title concept characters worldView plot referenceLinks writingRules]

      parsed.each_with_object({}) do |(key, value), result|
        result[key] = value.strip if allowed.include?(key) && value.is_a?(String)
      end
    end
  end
end
